%\VignetteIndexEntry{segmentsSeq: Methylation locus identification}
%\VignettePackage{segmentSeq}

\documentclass[a4paper]{article}

%\usepackage{rotating}

\title{segmentSeq: methods for detecting methylation loci and differential methylation}
\author{Thomas J. Hardcastle}

\RequirePackage{/home/bioinf/tjh48/R/x86_64-pc-linux-gnu-library/3.3/BiocStyle/resources/tex/Bioconductor}

\AtBeginDocument{\bibliographystyle{/home/bioinf/tjh48/R/x86_64-pc-linux-gnu-library/3.3/BiocStyle/resources/tex/unsrturl}}

\usepackage{Sweave}
\begin{document}

\maketitle

\section{Introduction}

This vignette introduces analysis methods for data from high-throughput sequencing of bisulphite treated DNA to detect cytosine methylation. The \verb'segmentSeq' package was originally designed to detect siRNA loci \cite{Hardcastle:2011} and many of the methods developed for this can be used to detect loci of cytosine methylation from replicated (or unreplicated) sequencing data.

\section{Preparation}

Preparation of the segmentSeq package proceeds as in siRNA analysis. We begin by loading the \verb'segmentSeq' package.

\begin{Schunk}
\begin{Sinput}
>   library(segmentSeq)
\end{Sinput}
\end{Schunk}

Note that because the experiments that \verb'segmentSeq' is designed to analyse are usually massive, we should use (if possible) parallel processing as implemented by the \verb'parallel' package. If using this approach, we need to begin by define a \textsl{cluster}. The following command will use eight processors on a single machine; see the help page for 'makeCluster' for more information. If we don't want to parallelise, we can proceed anyway with a \verb'NULL' cluster. Results may be slightly different depending on whether or not a cluster is used owing to the non-deterministic elements of the method.

\begin{Schunk}
\begin{Sinput}
> if(require("parallel")) 
+ {
+     numCores <- min(8, detectCores())
+     cl <- makeCluster(numCores)
+ } else {
+     cl <- NULL
+ }
\end{Sinput}
\end{Schunk}


The \verb'segmentSeq' package is designed to read in output from the YAMA (Yet Another Methylome Aligner) program. This is a perl-based package using either bowtie or bowtie2 to align bisulphite treated reads (in an unbiased manner) to a reference and identify the number of times each cytosine is identified as methylated or unmethylated. Unlike most other aligners, YAMA does not require that reads that map to more than one location are discarded, instead it reports the number of alternate matches to the reference for each cytosine. This is then used by \verb'segmentSeq' to weight the observed number of methylated/un-methylated cytosines at a location. The files used here have been compressed to save space.

\begin{Schunk}
\begin{Sinput}
> datadir <- system.file("extdata", package = "segmentSeq")
> files <- c("short_18B_C24_C24_trim.fastq_CG_methCalls.gz",
+ "short_Sample_17A_trimmed.fastq_CG_methCalls.gz",
+ "short_13_C24_col_trim.fastq_CG_methCalls.gz",
+ "short_Sample_28_trimmed.fastq_CG_methCalls.gz")
> mD <- readMeths(files = files, dir = datadir,
+ libnames = c("A1", "A2", "B1", "B2"), replicates = c("A","A","B","B"),
+ nonconversion = c(0.004777, 0.005903, 0.016514, 0.006134))
\end{Sinput}
\end{Schunk}

We can begin by plotting the distribution of methylation for these samples. The distribution can be plotted for each sample individually, or as an average across multiple samples. We can also subtract one distribution from another to visualise patterns of differential methylation on the genome.

\begin{Schunk}
\begin{Sinput}
> par(mfrow = c(2,1))
> dists <- plotMethDistribution(mD, main = "Distributions of methylation", chr = "Chr1")
> plotMethDistribution(mD, 
+                      subtract = rowMeans(sapply(dists, function(x) x[,2])), 
+                      main = "Differences between distributions", chr = "Chr1")
\end{Sinput}
\end{Schunk}

%\begin{sidewaysfigure}[!ht]
\begin{figure}[!ht]
\begin{center}
\includegraphics{methylationAnalysis-figMethDist}
\caption{Distributions of methylation on the genome (first two million bases of chromosome 1.}
\label{fig:Seg}
\end{center}
\end{figure}
%\end{sidewaysfigure}


Next, we process this \verb'alignmentData' object to produce a \verb'segData' object. This \verb'segData' object contains a set of potential segments on the genome defined by the start and end points of regions of overlapping alignments in the \verb'alignmentData' object. It then evaluates the number of tags that hit in each of these segments.

\begin{Schunk}
\begin{Sinput}
> sD <- processAD(mD, cl = cl)
\end{Sinput}
\end{Schunk}


We can now construct a segment map from these potential segments.

\subsection*{Segmentation by heuristic Bayesian methods}

A fast method of segmentation can be achieved by assuming a binomial distribution on the data with an uninformative beta prior, and identifying those potential segments which have a sufficiently large posterior likelihood that the proportion of methylation exceeds some critical value. This value may be determined by examining the data using the 'thresholdFinder' function, or supplied manually.

\begin{Schunk}
\begin{Sinput}
> thresh <- thresholdFinder("beta", mD, cl = cl, bootstrap = 1)
> hS <- heuristicSeg(sD = sD, aD = mD, prop = thresh, cl = cl, gap = 100, getLikes = FALSE)
> hS
\end{Sinput}
\begin{Soutput}
GRanges object with 4298 ranges and 0 metadata columns:
         seqnames             ranges strand
            <Rle>          <IRanges>  <Rle>
     [1]     Chr1         [108, 115]      *
     [2]     Chr1         [161, 500]      *
     [3]     Chr1         [511, 511]      *
     [4]     Chr1         [642, 650]      *
     [5]     Chr1         [809, 809]      *
     ...      ...                ...    ...
  [4294]     Chr1 [1988245, 1988245]      *
  [4295]     Chr1 [1990160, 1990160]      *
  [4296]     Chr1 [1990186, 1990298]      *
  [4297]     Chr1 [1993149, 1993149]      *
  [4298]     Chr1 [1993335, 1993335]      *
  -------
  seqinfo: 1 sequence from an unspecified genome; no seqlengths
An object of class "lociData"
4298 rows and 4 columns

Slot "replicates"
A A B B
Slot "groups":
list()

Slot "data":
     [,1]  [,2]  [,3]  [,4]
[1,] 64:12 14:6  31:9  1:1 
[2,] 49:13 47:21 40:15 8:3 
[3,] 10:0  9:2   8:2   5:0 
[4,] 55:18 75:31 36:9  10:2
[5,] 0:0   0:3   5:0   0:0 
4293 more rows...

Slot "annotation":
data frame with 0 columns and 4298 rows

Slot "locLikelihoods" (stored on log scale):
Matrix with  4298  rows.
       A   B
1      1   1
2      1   1
3      1   1
4      1   1
5      0   1
...  ... ...
4294   1   0
4295   0   1
4296   0   1
4297   1   1
4298   1   0

Expected number of loci in each replicate group
   A    B 
3199 3083 
\end{Soutput}
\end{Schunk}



Within a methylation locus, it is not uncommon to find completely unmethylated cytosines. If the coverage of these cytosines is too high, it is possible that these will cause the locus to be split into two or more fragments. The \verb'mergeMethSegs' function can be used to overcome this splitting by merging loci with identical patterns of expression that are not separated by too great a gap. Merging in this manner is optional, but recommended.

\begin{Schunk}
\begin{Sinput}
> hS <- mergeMethSegs(hS, mD, gap = 5000, cl = cl)
\end{Sinput}
\end{Schunk}

We can then estimate posterior likelihoods on the defined loci by applying empirical Bayesian methods. These will not change the locus definition, but will assign likelihoods that the identified loci represent a true methylation locus in each replicate group.

\begin{Schunk}
\begin{Sinput}
> hSL <- lociLikelihoods(hS, mD, cl = cl)
\end{Sinput}
\begin{Soutput}
......
\end{Soutput}
\end{Schunk}

